Problems when following a non-main goroutine with 'next' command: Delve doesn't detect the end of the program

Playing with a basic [goroutine example](https://gobyexample.com/goroutines) to add it to the getting started guide I came across a series of panics and hangs (that I believe are all related) when switching to a non-`main` goroutine and following it with `next`, because (I'm guessing) Delve at some point misses the end of the program and pointers like `(*Thread)` in the server or `DebuggerState.CurrentThread` in the client are `nil`. I'll be looking deeper into this issue but I document it here to get some feedback about it.

The problem, as it deals with threads and goroutines, is not completely deterministic but comes up often enough (at least in my machine) that after a couple of tries it will manifest itself (in different ways). In it most basic form the test code is an empty main function that spawns a goroutine (also empty) and waits for it (the `sync.Wait` was added to the cited example):

```go
package main

import (
    "sync"
)

var wg sync.WaitGroup

func f(from string) {
    defer wg.Done()
    return
}

func main() {
    wg.Add(1)
    go f("goroutine")
    wg.Wait()
}
```

Building the binary

```
$ dlv debug --output debug_binary
```

and running the following CLI commands that break inside the function executed by the goroutine (with an implicit goroutine switch)

```
r
b main.main
c
b f
c
n # repeated three times to exit f() goroutine
n
n
```

leaves it inside the created goroutine

```
     8:	
=>   9:	func f(from string) {
    10:	    defer wg.Done()
    11:	    return
    12:	}
    13:	
    14:	func main() {
(dlv) goroutines
[4 goroutines]
  Goroutine 1 - User: /usr/local/go/src/runtime/sema.go:56 sync.runtime_Semacquire (0x4350b9)
  Goroutine 2 - User: /usr/local/go/src/runtime/proc.go:288 runtime.gopark (0x427b2d)
  Goroutine 17 - User: /usr/local/go/src/runtime/proc.go:288 runtime.gopark (0x427b2d)
* Goroutine 18 - User: ./main.go:9 main.f (0x455f1f) (thread 21588)
(dlv) 
```

From this point, following the execution with the `next` command sometimes ends correctly with the process termination, but in other cases it results in panics or hangs. Alternatively, using the command `continue` always ends correctly without issues.

Problem 1: panic in the terminal client.

```
No current thread available
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x7be5fb]

goroutine 1 [running]:
github.com/derekparker/delve/pkg/terminal.continueUntilCompleteNext(0xc420061e60, 0xc4204362a0, 0xaecfb2, 0x4, 0x0, 0x0)
	/home/user/go/src/github.com/derekparker/delve/pkg/terminal/command.go:749 +0x20b
github.com/derekparker/delve/pkg/terminal.next(0xc420061e60, 0x0, 0xffffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x2, 0x0)
	/home/user/go/src/github.com/derekparker/delve/pkg/terminal/command.go:823 +0x116
github.com/derekparker/delve/pkg/terminal.(*Commands).CallWithContext(0xc420171d40, 0xc420293bec, 0x0, 0xc420061e60, 0x0, 0xffffffffffffffff, 0x0, 0x0, 0x0, 0x0)
	/home/user/go/src/github.com/derekparker/delve/pkg/terminal/command.go:348 +0x12a
github.com/derekparker/delve/pkg/terminal.(*Commands).Call(0xc420171d40, 0xc420293bec, 0x0, 0xc420061e60, 0x0, 0x0)
	/home/user/go/src/github.com/derekparker/delve/pkg/terminal/command.go:353 +0x71
github.com/derekparker/delve/pkg/terminal.(*Term).Run(0xc420061e60, 0x0, 0x0, 0x0)
	/home/user/go/src/github.com/derekparker/delve/pkg/terminal/terminal.go:132 +0x33c
github.com/derekparker/delve/cmd/dlv/cmds.execute(0x0, 0xc42014e3b0, 0x1, 0x1, 0xc420156240, 0x0, 0x0, 0x1, 0x0)
	/home/user/go/src/github.com/derekparker/delve/cmd/dlv/cmds/commands.go:537 +0x8e3
github.com/derekparker/delve/cmd/dlv/cmds.debugCmd.func1(0xc420172900, 0xc42014c480, 0x0, 0x2, 0x0)
	/home/user/go/src/github.com/derekparker/delve/cmd/dlv/cmds/commands.go:300 +0x405
github.com/derekparker/delve/cmd/dlv/cmds.debugCmd(0xc420172900, 0xc42014c480, 0x0, 0x2)
	/home/user/go/src/github.com/derekparker/delve/cmd/dlv/cmds/commands.go:301 +0x49
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).execute(0xc420172900, 0xc42014c300, 0x2, 0x2, 0x0, 0x0)
	/home/user/go/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:647 +0xa4e
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420172240, 0xc420172900, 0x0, 0x0)
	/home/user/go/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:733 +0x68c
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420172240, 0x0, 0x0)
	/home/user/go/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:692 +0x4f
main.main()
	/home/user/go/src/github.com/derekparker/delve/cmd/dlv/main.go:18 +0x91

```

Problem 2: panic in the debugger server when restarting what appeared to be a correct end in the execution

```
(dlv) r
Command failed: Internal debugger error: runtime error: invalid memory address or nil pointer dereference
runtime.call32 (0x45ceba)
	/usr/local/go/src/runtime/asm_amd64.s:509
runtime.gopanic (0x42f541)
	/usr/local/go/src/runtime/panic.go:491
runtime.panicmem (0x42e41d)
	/usr/local/go/src/runtime/panic.go:63
runtime.sigpanic (0x44649b)
	/usr/local/go/src/runtime/signal_unix.go:367
github.com/derekparker/delve/pkg/proc/native.(*Thread).stopped (0x797392)
	/home/user/go/src/github.com/derekparker/delve/pkg/proc/native/threads_linux.go:34
github.com/derekparker/delve/pkg/proc/native.(*Thread).Stopped (0x796a8a)
	/home/user/go/src/github.com/derekparker/delve/pkg/proc/native/threads.go:120
github.com/derekparker/delve/pkg/proc/native.(*Process).Kill (0x7916fd)
	/home/user/go/src/github.com/derekparker/delve/pkg/proc/native/proc_linux.go:106
github.com/derekparker/delve/pkg/proc/native.(*Process).Detach (0x78ee2e)
	/home/user/go/src/github.com/derekparker/delve/pkg/proc/native/proc.go:82
github.com/derekparker/delve/service/debugger.(*Debugger).detach (0x79a61b)
	/home/user/go/src/github.com/derekparker/delve/service/debugger/debugger.go:177
github.com/derekparker/delve/service/debugger.(*Debugger).Restart (0x79a78d)
	/home/user/go/src/github.com/derekparker/delve/service/debugger/debugger.go:203
github.com/derekparker/delve/service/rpc2.(*RPCServer).Restart (0x977f87)
	/home/user/go/src/github.com/derekparker/delve/service/rpc2/server.go:83
runtime.call128 (0x45cfd1)
	/usr/local/go/src/runtime/asm_amd64.s:511
reflect.Value.call (0x4e21b5)
	/usr/local/go/src/reflect/value.go:434
reflect.Value.Call (0x4e1366)
	/usr/local/go/src/reflect/value.go:302
github.com/derekparker/delve/service/rpccommon.(*ServerImpl).serveJSONCodec.func1 (0x98095d)
	/home/user/go/src/github.com/derekparker/delve/service/rpccommon/server.go:299
github.com/derekparker/delve/service/rpccommon.(*ServerImpl).serveJSONCodec (0x97f52d)
	/home/user/go/src/github.com/derekparker/delve/service/rpccommon/server.go:301
runtime.goexit (0x45f6e0)
	/usr/local/go/src/runtime/asm_amd64.s:2337
```

Problem 3: Delve hangs with a zombie process.

```
$ command ps aux -T | grep debug_bin
user     15052 15052  0.0  0.0      0     0 pts/18   Zl   08:58   0:00 [debug_binary] <defunct>
user     15052 15053  0.0  0.0      0     0 pts/18   Zl   08:58   0:00 [debug_binary] <defunct>
user     15052 15054  0.0  0.0      0     0 pts/18   Zl   08:58   0:00 [debug_binary] <defunct>
```

Removing the `Wait` call in `main`, although it makes it less probable that Delve will hit the breakpoint inside the goroutine (because `main` ends before that), when it does, it always (or very likely) causes one of the above described problems because (I'm guessing again) Delve is not detecting the end of the `main` goroutine while stepping through the (now dead) non-`main` goroutine. (To increase the probability of hitting the breakpoint the removed `Println` statement from the cited example should be added again to the function executed by the goroutine.)

Development environment:

```
go version go1.9.2 linux/amd64
Delve Debugger
Version: 1.0.0-rc.2
Build: 74d330a013a5412649aaa069d932b93889c2dbd2
```

--- a/_fixtures/issue1101.go
+++ b/_fixtures/issue1101.go
@@ -1,0 +1,20 @@
+package main
+
+import (
+	"os"
+	"sync"
+)
+
+var wg sync.WaitGroup
+
+func f(from string) {
+	defer wg.Done()
+	return
+}
+
+func main() {
+	wg.Add(1)
+	go f("goroutine")
+	wg.Wait()
+	os.Exit(2)
+}

--- a/pkg/proc/core/core.go
+++ b/pkg/proc/core/core.go
@@ -302,10 +302,6 @@ func (p *Process) AllGCache() *[]*proc.G {
 	return &p.allGCache
 }
 
-func (p *Process) Halt() error {
-	return nil
-}
-
 func (p *Process) Pid() int {
 	return p.core.Pid
 }

--- a/pkg/proc/core/core.go
+++ b/pkg/proc/core/core.go
@@ -306,10 +306,6 @@ func (p *Process) Halt() error {
 	return nil
 }
 
-func (p *Process) Kill() error {
-	return nil
-}
-
 func (p *Process) Pid() int {
 	return p.core.Pid
 }

--- a/pkg/proc/gdbserial/gdbserver.go
+++ b/pkg/proc/gdbserial/gdbserver.go
@@ -745,24 +745,14 @@ func (p *Process) Halt() error {
 	return p.conn.sendCtrlC()
 }
 
-func (p *Process) Kill() error {
-	if p.exited {
-		return nil
-	}
-	err := p.conn.kill()
-	if _, exited := err.(proc.ProcessExitedError); exited {
-		p.exited = true
-		return nil
-	}
-	return err
-}
-
 func (p *Process) Detach(kill bool) error {
-	if kill {
-		if err := p.Kill(); err != nil {
+	if kill && !p.exited {
+		err := p.conn.kill()
+		if err != nil {
 			if _, exited := err.(proc.ProcessExitedError); !exited {
 				return err
 			}
+			p.exited = true
 		}
 	}
 	if !p.exited {

--- a/pkg/proc/gdbserial/gdbserver.go
+++ b/pkg/proc/gdbserial/gdbserver.go
@@ -737,14 +737,6 @@ func (p *Process) getCtrlC() bool {
 	return p.ctrlC
 }
 
-func (p *Process) Halt() error {
-	if p.exited {
-		return nil
-	}
-	p.setCtrlC(true)
-	return p.conn.sendCtrlC()
-}
-
 func (p *Process) Detach(kill bool) error {
 	if kill && !p.exited {
 		err := p.conn.kill()

--- a/pkg/proc/gdbserial/rr_test.go
+++ b/pkg/proc/gdbserial/rr_test.go
@@ -26,7 +26,6 @@ func withTestRecording(name string, t testing.TB, fn func(p *gdbserial.Process, 
 	t.Logf("replaying %q", tracedir)
 
 	defer func() {
-		p.Halt()
 		p.Detach(true)
 		if tracedir != "" {
 			protest.SafeRemoveAll(tracedir)

--- a/pkg/proc/interface.go
+++ b/pkg/proc/interface.go
@@ -90,7 +90,6 @@ type ProcessManipulation interface {
 	// CheckAndClearManualStopRequest returns true the first time it's called
 	// after a call to RequestManualStop.
 	CheckAndClearManualStopRequest() bool
-	Halt() error
 	Detach(bool) error
 }
 

--- a/pkg/proc/interface.go
+++ b/pkg/proc/interface.go
@@ -91,7 +91,6 @@ type ProcessManipulation interface {
 	// after a call to RequestManualStop.
 	CheckAndClearManualStopRequest() bool
 	Halt() error
-	Kill() error
 	Detach(bool) error
 }
 

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -246,10 +246,7 @@ func (dbp *Process) ContinueOnce() (proc.Thread, error) {
 	if err != nil {
 		return nil, err
 	}
-	if err := dbp.Halt(); err != nil {
-		return nil, dbp.exitGuard(err)
-	}
-	if err := dbp.setCurrentBreakpoints(trapthread); err != nil {
+	if err := dbp.stop(trapthread); err != nil {
 		return nil, err
 	}
 	return trapthread, err
@@ -321,19 +318,6 @@ func (dbp *Process) SwitchGoroutine(gid int) error {
 		return dbp.SwitchThread(g.Thread.ThreadID())
 	}
 	dbp.selectedGoroutine = g
-	return nil
-}
-
-// Halt stops all threads.
-func (dbp *Process) Halt() (err error) {
-	if dbp.exited {
-		return &proc.ProcessExitedError{Pid: dbp.Pid()}
-	}
-	for _, th := range dbp.threads {
-		if err := th.Halt(); err != nil {
-			return err
-		}
-	}
 	return nil
 }
 

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -79,7 +79,7 @@ func (dbp *Process) Detach(kill bool) (err error) {
 		return nil
 	}
 	if kill && dbp.childProcess {
-		err := dbp.Kill()
+		err := dbp.kill()
 		if err != nil {
 			return err
 		}

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -32,8 +32,7 @@ type Process struct {
 	allGCache           []*proc.G
 	os                  *OSProcessDetails
 	firstStart          bool
-	haltMu              sync.Mutex
-	halt                bool
+	stopMu              sync.Mutex
 	resumeChan          chan<- struct{}
 	exited              bool
 	ptraceChan          chan func()
@@ -176,18 +175,17 @@ func (dbp *Process) RequestManualStop() error {
 	if dbp.exited {
 		return &proc.ProcessExitedError{Pid: dbp.Pid()}
 	}
-	dbp.haltMu.Lock()
-	defer dbp.haltMu.Unlock()
+	dbp.stopMu.Lock()
+	defer dbp.stopMu.Unlock()
 	dbp.manualStopRequested = true
-	dbp.halt = true
 	return dbp.requestManualStop()
 }
 
 func (dbp *Process) CheckAndClearManualStopRequest() bool {
-	dbp.haltMu.Lock()
+	dbp.stopMu.Lock()
 	msr := dbp.manualStopRequested
 	dbp.manualStopRequested = false
-	dbp.haltMu.Unlock()
+	dbp.stopMu.Unlock()
 	return msr
 }
 

--- a/pkg/proc/native/proc_darwin.go
+++ b/pkg/proc/native/proc_darwin.go
@@ -166,7 +166,7 @@ func Attach(pid int) (*Process, error) {
 }
 
 // Kill kills the process.
-func (dbp *Process) Kill() (err error) {
+func (dbp *Process) kill() (err error) {
 	if dbp.exited {
 		return nil
 	}

--- a/pkg/proc/native/proc_darwin.go
+++ b/pkg/proc/native/proc_darwin.go
@@ -111,12 +111,7 @@ func Launch(cmd []string, wd string) (*Process, error) {
 	if err != nil {
 		return nil, err
 	}
-	if err := dbp.Halt(); err != nil {
-		return nil, dbp.exitGuard(err)
-	}
-
-	_, err = dbp.waitForStop()
-	if err != nil {
+	if err := dbp.stop(nil); err != nil {
 		return nil, err
 	}
 
@@ -384,23 +379,6 @@ func (dbp *Process) waitForStop() ([]int, error) {
 	}
 }
 
-func (dbp *Process) setCurrentBreakpoints(trapthread *Thread) error {
-	ports, err := dbp.waitForStop()
-	if err != nil {
-		return err
-	}
-	trapthread.SetCurrentBreakpoint()
-	for _, port := range ports {
-		if th, ok := dbp.threads[port]; ok {
-			err := th.SetCurrentBreakpoint()
-			if err != nil {
-				return err
-			}
-		}
-	}
-	return nil
-}
-
 func (dbp *Process) loadProcessInformation(wg *sync.WaitGroup) {
 	wg.Done()
 }
@@ -441,6 +419,39 @@ func (dbp *Process) resume() error {
 	for _, thread := range dbp.threads {
 		if err := thread.resume(); err != nil {
 			return dbp.exitGuard(err)
+		}
+	}
+	return nil
+}
+
+// stop stops all running threads threads and sets breakpoints
+func (dbp *Process) stop(trapthread *Thread) (err error) {
+	if dbp.exited {
+		return &proc.ProcessExitedError{Pid: dbp.Pid()}
+	}
+	for _, th := range dbp.threads {
+		if !th.Stopped() {
+			if err := th.halt(); err != nil {
+				return dbp.exitGuard(err)
+			}
+		}
+		th.running = false
+	}
+
+	ports, err := dbp.waitForStop()
+	if err != nil {
+		return err
+	}
+	if !dbp.os.initialized {
+		return nil
+	}
+	trapthread.SetCurrentBreakpoint()
+	for _, port := range ports {
+		if th, ok := dbp.threads[port]; ok {
+			err := th.SetCurrentBreakpoint()
+			if err != nil {
+				return err
+			}
 		}
 	}
 	return nil

--- a/pkg/proc/native/proc_darwin.go
+++ b/pkg/proc/native/proc_darwin.go
@@ -25,6 +25,7 @@ type OSProcessDetails struct {
 	exceptionPort    C.mach_port_t // mach port for receiving mach exceptions.
 	notificationPort C.mach_port_t // mach port for dead name notification (process exit).
 	initialized      bool
+	halt             bool
 
 	// the main port we use, will return messages from both the
 	// exception and notification ports.
@@ -191,6 +192,7 @@ func (dbp *Process) requestManualStop() (err error) {
 		thread        = C.mach_port_t(dbp.currentThread.os.threadAct)
 		exceptionPort = C.mach_port_t(dbp.os.exceptionPort)
 	)
+	dbp.os.halt = true
 	kret := C.raise_exception(task, thread, exceptionPort, C.EXC_BREAKPOINT)
 	if kret != C.KERN_SUCCESS {
 		return fmt.Errorf("could not raise mach exception")
@@ -306,9 +308,9 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 			return nil, proc.ProcessExitedError{Pid: dbp.pid, Status: status.ExitStatus()}
 
 		case C.MACH_RCV_INTERRUPTED:
-			dbp.haltMu.Lock()
-			halt := dbp.halt
-			dbp.haltMu.Unlock()
+			dbp.stopMu.Lock()
+			halt := dbp.os.halt
+			dbp.stopMu.Unlock()
 			if !halt {
 				// Call trapWait again, it seems
 				// MACH_RCV_INTERRUPTED is emitted before
@@ -337,11 +339,11 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 		dbp.updateThreadList()
 		th, ok := dbp.threads[int(port)]
 		if !ok {
-			dbp.haltMu.Lock()
-			halt := dbp.halt
-			dbp.haltMu.Unlock()
+			dbp.stopMu.Lock()
+			halt := dbp.os.halt
+			dbp.stopMu.Unlock()
 			if halt {
-				dbp.halt = false
+				dbp.os.halt = false
 				return th, nil
 			}
 			if dbp.firstStart || th.singleStepping {
@@ -424,14 +426,14 @@ func (dbp *Process) resume() error {
 	return nil
 }
 
-// stop stops all running threads threads and sets breakpoints
+// stop stops all running threads and sets breakpoints
 func (dbp *Process) stop(trapthread *Thread) (err error) {
 	if dbp.exited {
 		return &proc.ProcessExitedError{Pid: dbp.Pid()}
 	}
 	for _, th := range dbp.threads {
 		if !th.Stopped() {
-			if err := th.halt(); err != nil {
+			if err := th.stop(); err != nil {
 				return dbp.exitGuard(err)
 			}
 		}

--- a/pkg/proc/native/proc_darwin.go
+++ b/pkg/proc/native/proc_darwin.go
@@ -435,7 +435,6 @@ func (dbp *Process) stop(trapthread *Thread) (err error) {
 				return dbp.exitGuard(err)
 			}
 		}
-		th.running = false
 	}
 
 	ports, err := dbp.waitForStop()

--- a/pkg/proc/native/proc_linux.go
+++ b/pkg/proc/native/proc_linux.go
@@ -240,8 +240,8 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 			dbp.haltMu.Unlock()
 			if halt {
 				dbp.halt = false
-				th.running = false
-				dbp.threads[int(wpid)].running = false
+				th.os.running = false
+				dbp.threads[int(wpid)].os.running = false
 				return nil, nil
 			}
 			if err = th.Continue(); err != nil {
@@ -267,12 +267,12 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 		halt := dbp.halt
 		dbp.haltMu.Unlock()
 		if halt && (status.StopSignal() == sys.SIGTRAP || status.StopSignal() == sys.SIGSTOP) {
-			th.running = false
+			th.os.running = false
 			dbp.halt = false
 			return th, nil
 		}
 		if status.StopSignal() == sys.SIGTRAP {
-			th.running = false
+			th.os.running = false
 			return th, nil
 		}
 		if th != nil {
@@ -425,7 +425,7 @@ func (dbp *Process) stop(trapthread *Thread) (err error) {
 	for {
 		allstopped := true
 		for _, th := range dbp.threads {
-			if th.running {
+			if th.os.running {
 				allstopped = false
 				break
 			}

--- a/pkg/proc/native/proc_linux.go
+++ b/pkg/proc/native/proc_linux.go
@@ -98,8 +98,8 @@ func Attach(pid int) (*Process, error) {
 	return dbp, nil
 }
 
-// Kill kills the target process.
-func (dbp *Process) Kill() (err error) {
+// kill kills the target process.
+func (dbp *Process) kill() (err error) {
 	if dbp.exited {
 		return nil
 	}

--- a/pkg/proc/native/proc_linux.go
+++ b/pkg/proc/native/proc_linux.go
@@ -230,9 +230,19 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 			if err != nil {
 				if err == sys.ESRCH {
 					// thread died while we were adding it
+					delete(dbp.threads, int(cloned))
 					continue
 				}
 				return nil, err
+			}
+			dbp.haltMu.Lock()
+			halt := dbp.halt
+			dbp.haltMu.Unlock()
+			if halt {
+				dbp.halt = false
+				th.running = false
+				dbp.threads[int(wpid)].running = false
+				return nil, nil
 			}
 			if err = th.Continue(); err != nil {
 				if err == sys.ESRCH {
@@ -256,7 +266,7 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 		dbp.haltMu.Lock()
 		halt := dbp.halt
 		dbp.haltMu.Unlock()
-		if status.StopSignal() == sys.SIGTRAP && halt {
+		if halt && (status.StopSignal() == sys.SIGTRAP || status.StopSignal() == sys.SIGSTOP) {
 			th.running = false
 			dbp.halt = false
 			return th, nil
@@ -368,16 +378,28 @@ func (dbp *Process) wait(pid, options int) (int, *sys.WaitStatus, error) {
 }
 
 func (dbp *Process) setCurrentBreakpoints(trapthread *Thread) error {
+	// wait for all threads to stop
+	for {
+		allstopped := true
+		for _, th := range dbp.threads {
+			if th.running {
+				allstopped = false
+				break
+			}
+		}
+		if allstopped {
+			break
+		}
+		dbp.halt = true
+		_, err := dbp.trapWait(-1)
+		if err != nil {
+			return err
+		}
+	}
 	for _, th := range dbp.threads {
 		if th.CurrentBreakpoint.Breakpoint == nil {
 			if err := th.SetCurrentBreakpoint(); err != nil {
-				if err == sys.ESRCH {
-					// This thread quit between the point where we received the breakpoint and
-					// the stop signal.
-					delete(dbp.threads, th.ID)
-				} else {
-					return err
-				}
+				return err
 			}
 		}
 	}

--- a/pkg/proc/native/proc_linux.go
+++ b/pkg/proc/native/proc_linux.go
@@ -194,6 +194,10 @@ func findExecutable(path string, pid int) string {
 }
 
 func (dbp *Process) trapWait(pid int) (*Thread, error) {
+	return dbp.trapWaitInternal(pid, false)
+}
+
+func (dbp *Process) trapWaitInternal(pid int, halt bool) (*Thread, error) {
 	for {
 		wpid, status, err := dbp.wait(pid, 0)
 		if err != nil {
@@ -235,11 +239,7 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 				}
 				return nil, err
 			}
-			dbp.haltMu.Lock()
-			halt := dbp.halt
-			dbp.haltMu.Unlock()
 			if halt {
-				dbp.halt = false
 				th.os.running = false
 				dbp.threads[int(wpid)].os.running = false
 				return nil, nil
@@ -263,15 +263,7 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 			// Sometimes we get an unknown thread, ignore it?
 			continue
 		}
-		dbp.haltMu.Lock()
-		halt := dbp.halt
-		dbp.haltMu.Unlock()
-		if halt && (status.StopSignal() == sys.SIGTRAP || status.StopSignal() == sys.SIGSTOP) {
-			th.os.running = false
-			dbp.halt = false
-			return th, nil
-		}
-		if status.StopSignal() == sys.SIGTRAP {
+		if (halt && status.StopSignal() == sys.SIGSTOP) || (status.StopSignal() == sys.SIGTRAP) {
 			th.os.running = false
 			return th, nil
 		}
@@ -382,7 +374,7 @@ func (dbp *Process) exitGuard(err error) error {
 		return err
 	}
 	if status(dbp.pid, dbp.os.comm) == StatusZombie {
-		_, err := dbp.trapWait(-1)
+		_, err := dbp.trapWaitInternal(-1, false)
 		return err
 	}
 
@@ -415,7 +407,7 @@ func (dbp *Process) stop(trapthread *Thread) (err error) {
 	}
 	for _, th := range dbp.threads {
 		if !th.Stopped() {
-			if err := th.halt(); err != nil {
+			if err := th.stop(); err != nil {
 				return dbp.exitGuard(err)
 			}
 		}
@@ -433,8 +425,7 @@ func (dbp *Process) stop(trapthread *Thread) (err error) {
 		if allstopped {
 			break
 		}
-		dbp.halt = true
-		_, err := dbp.trapWait(-1)
+		_, err := dbp.trapWaitInternal(-1, true)
 		if err != nil {
 			return err
 		}

--- a/pkg/proc/native/proc_linux.go
+++ b/pkg/proc/native/proc_linux.go
@@ -377,35 +377,6 @@ func (dbp *Process) wait(pid, options int) (int, *sys.WaitStatus, error) {
 	}
 }
 
-func (dbp *Process) setCurrentBreakpoints(trapthread *Thread) error {
-	// wait for all threads to stop
-	for {
-		allstopped := true
-		for _, th := range dbp.threads {
-			if th.running {
-				allstopped = false
-				break
-			}
-		}
-		if allstopped {
-			break
-		}
-		dbp.halt = true
-		_, err := dbp.trapWait(-1)
-		if err != nil {
-			return err
-		}
-	}
-	for _, th := range dbp.threads {
-		if th.CurrentBreakpoint.Breakpoint == nil {
-			if err := th.SetCurrentBreakpoint(); err != nil {
-				return err
-			}
-		}
-	}
-	return nil
-}
-
 func (dbp *Process) exitGuard(err error) error {
 	if err != sys.ESRCH {
 		return err
@@ -432,6 +403,49 @@ func (dbp *Process) resume() error {
 	for _, thread := range dbp.threads {
 		if err := thread.resume(); err != nil && err != sys.ESRCH {
 			return err
+		}
+	}
+	return nil
+}
+
+// stop stops all running threads threads and sets breakpoints
+func (dbp *Process) stop(trapthread *Thread) (err error) {
+	if dbp.exited {
+		return &proc.ProcessExitedError{Pid: dbp.Pid()}
+	}
+	for _, th := range dbp.threads {
+		if !th.Stopped() {
+			if err := th.halt(); err != nil {
+				return dbp.exitGuard(err)
+			}
+		}
+	}
+
+	// wait for all threads to stop
+	for {
+		allstopped := true
+		for _, th := range dbp.threads {
+			if th.running {
+				allstopped = false
+				break
+			}
+		}
+		if allstopped {
+			break
+		}
+		dbp.halt = true
+		_, err := dbp.trapWait(-1)
+		if err != nil {
+			return err
+		}
+	}
+
+	// set breakpoints on all threads
+	for _, th := range dbp.threads {
+		if th.CurrentBreakpoint.Breakpoint == nil {
+			if err := th.SetCurrentBreakpoint(); err != nil {
+				return err
+			}
 		}
 	}
 	return nil

--- a/pkg/proc/native/proc_windows.go
+++ b/pkg/proc/native/proc_windows.go
@@ -2,7 +2,6 @@ package native
 
 import (
 	"debug/pe"
-	"errors"
 	"fmt"
 	"io"
 	"os"
@@ -174,9 +173,6 @@ func Attach(pid int) (*Process, error) {
 func (dbp *Process) kill() error {
 	if dbp.exited {
 		return nil
-	}
-	if !dbp.threads[dbp.pid].Stopped() {
-		return errors.New("process must be stopped in order to kill it")
 	}
 
 	p, err := os.FindProcess(dbp.pid)
@@ -393,7 +389,40 @@ func (dbp *Process) wait(pid, options int) (int, *sys.WaitStatus, error) {
 	return 0, nil, fmt.Errorf("not implemented: wait")
 }
 
-func (dbp *Process) setCurrentBreakpoints(trapthread *Thread) error {
+func (dbp *Process) exitGuard(err error) error {
+	return err
+}
+
+func (dbp *Process) resume() error {
+	for _, thread := range dbp.threads {
+		if thread.CurrentBreakpoint.Breakpoint != nil {
+			if err := thread.StepInstruction(); err != nil {
+				return err
+			}
+			thread.CurrentBreakpoint.Clear()
+		}
+	}
+
+	for _, thread := range dbp.threads {
+		thread.running = true
+		_, err := _ResumeThread(thread.os.hThread)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// stop stops all running threads threads and sets breakpoints
+func (dbp *Process) stop(trapthread *Thread) (err error) {
+	if dbp.exited {
+		return &proc.ProcessExitedError{Pid: dbp.Pid()}
+	}
+	for _, th := range dbp.threads {
+		th.running = false
+	}
+
 	// While the debug event that stopped the target was being propagated
 	// other target threads could generate other debug events.
 	// After this function we need to know about all the threads
@@ -404,7 +433,7 @@ func (dbp *Process) setCurrentBreakpoints(trapthread *Thread) error {
 	// call to _ContinueDebugEvent will resume execution of some of the
 	// target threads.
 
-	err := trapthread.SetCurrentBreakpoint()
+	err = trapthread.SetCurrentBreakpoint()
 	if err != nil {
 		return err
 	}
@@ -433,31 +462,6 @@ func (dbp *Process) setCurrentBreakpoints(trapthread *Thread) error {
 			break
 		}
 		err = dbp.threads[tid].SetCurrentBreakpoint()
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
-}
-
-func (dbp *Process) exitGuard(err error) error {
-	return err
-}
-
-func (dbp *Process) resume() error {
-	for _, thread := range dbp.threads {
-		if thread.CurrentBreakpoint.Breakpoint != nil {
-			if err := thread.StepInstruction(); err != nil {
-				return err
-			}
-			thread.CurrentBreakpoint.Clear()
-		}
-	}
-
-	for _, thread := range dbp.threads {
-		thread.running = true
-		_, err := _ResumeThread(thread.os.hThread)
 		if err != nil {
 			return err
 		}

--- a/pkg/proc/native/proc_windows.go
+++ b/pkg/proc/native/proc_windows.go
@@ -404,7 +404,6 @@ func (dbp *Process) resume() error {
 	}
 
 	for _, thread := range dbp.threads {
-		thread.running = true
 		_, err := _ResumeThread(thread.os.hThread)
 		if err != nil {
 			return err
@@ -418,9 +417,6 @@ func (dbp *Process) resume() error {
 func (dbp *Process) stop(trapthread *Thread) (err error) {
 	if dbp.exited {
 		return &proc.ProcessExitedError{Pid: dbp.Pid()}
-	}
-	for _, th := range dbp.threads {
-		th.running = false
 	}
 
 	// While the debug event that stopped the target was being propagated
@@ -439,7 +435,6 @@ func (dbp *Process) stop(trapthread *Thread) (err error) {
 	}
 
 	for _, thread := range dbp.threads {
-		thread.running = false
 		_, err := _SuspendThread(thread.os.hThread)
 		if err != nil {
 			return err

--- a/pkg/proc/native/proc_windows.go
+++ b/pkg/proc/native/proc_windows.go
@@ -170,8 +170,8 @@ func Attach(pid int) (*Process, error) {
 	return dbp, nil
 }
 
-// Kill kills the process.
-func (dbp *Process) Kill() error {
+// kill kills the process.
+func (dbp *Process) kill() error {
 	if dbp.exited {
 		return nil
 	}

--- a/pkg/proc/native/threads.go
+++ b/pkg/proc/native/threads.go
@@ -113,13 +113,6 @@ func (thread *Thread) SetPC(pc uint64) error {
 	return regs.SetPC(thread, pc)
 }
 
-// Stopped returns whether the thread is stopped at
-// the operating system level. Actual implementation
-// is OS dependant, look in OS thread file.
-func (thread *Thread) Stopped() bool {
-	return thread.stopped()
-}
-
 // SetCurrentBreakpoint sets the current breakpoint that this
 // thread is stopped at as CurrentBreakpoint on the thread struct.
 func (thread *Thread) SetCurrentBreakpoint() error {

--- a/pkg/proc/native/threads.go
+++ b/pkg/proc/native/threads.go
@@ -120,22 +120,6 @@ func (thread *Thread) Stopped() bool {
 	return thread.stopped()
 }
 
-// Halt stops this thread from executing. Actual
-// implementation is OS dependant. Look in OS
-// thread file.
-func (thread *Thread) Halt() (err error) {
-	defer func() {
-		if err == nil {
-			thread.running = false
-		}
-	}()
-	if thread.Stopped() {
-		return
-	}
-	err = thread.halt()
-	return
-}
-
 // SetCurrentBreakpoint sets the current breakpoint that this
 // thread is stopped at as CurrentBreakpoint on the thread struct.
 func (thread *Thread) SetCurrentBreakpoint() error {

--- a/pkg/proc/native/threads.go
+++ b/pkg/proc/native/threads.go
@@ -18,7 +18,6 @@ type Thread struct {
 
 	dbp            *Process
 	singleStepping bool
-	running        bool
 	os             *OSSpecificDetails
 }
 
@@ -49,11 +48,9 @@ func (thread *Thread) Continue() error {
 // execute the instruction, and then replace the breakpoint.
 // Otherwise we simply execute the next instruction.
 func (thread *Thread) StepInstruction() (err error) {
-	thread.running = true
 	thread.singleStepping = true
 	defer func() {
 		thread.singleStepping = false
-		thread.running = false
 	}()
 	pc, err := thread.PC()
 	if err != nil {

--- a/pkg/proc/native/threads_darwin.go
+++ b/pkg/proc/native/threads_darwin.go
@@ -27,20 +27,6 @@ type OSSpecificDetails struct {
 // be continued.
 var ErrContinueThread = fmt.Errorf("could not continue thread")
 
-// Halt stops this thread from executing.
-func (thread *Thread) Halt() (err error) {
-	defer func() {
-		if err == nil {
-			thread.running = false
-		}
-	}()
-	if thread.Stopped() {
-		return
-	}
-	err = thread.halt()
-	return
-}
-
 func (t *Thread) halt() (err error) {
 	kret := C.thread_suspend(t.os.threadAct)
 	if kret != C.KERN_SUCCESS {
@@ -116,7 +102,9 @@ func (t *Thread) Blocked() bool {
 	}
 }
 
-func (t *Thread) stopped() bool {
+// Stopped returns whether the thread is stopped at
+// the operating system level.
+func (t *Thread) Stopped() bool {
 	return C.thread_blocked(t.os.threadAct) > C.int(0)
 }
 

--- a/pkg/proc/native/threads_darwin.go
+++ b/pkg/proc/native/threads_darwin.go
@@ -69,7 +69,6 @@ func (t *Thread) singleStep() error {
 }
 
 func (t *Thread) resume() error {
-	t.running = true
 	// TODO(dp) set flag for ptrace stops
 	var err error
 	t.dbp.execPtraceFunc(func() { err = PtraceCont(t.dbp.pid, 0) })

--- a/pkg/proc/native/threads_darwin.go
+++ b/pkg/proc/native/threads_darwin.go
@@ -27,7 +27,7 @@ type OSSpecificDetails struct {
 // be continued.
 var ErrContinueThread = fmt.Errorf("could not continue thread")
 
-func (t *Thread) halt() (err error) {
+func (t *Thread) stop() (err error) {
 	kret := C.thread_suspend(t.os.threadAct)
 	if kret != C.KERN_SUCCESS {
 		errStr := C.GoString(C.mach_error_string(C.mach_error_t(kret)))

--- a/pkg/proc/native/threads_darwin.go
+++ b/pkg/proc/native/threads_darwin.go
@@ -27,6 +27,20 @@ type OSSpecificDetails struct {
 // be continued.
 var ErrContinueThread = fmt.Errorf("could not continue thread")
 
+// Halt stops this thread from executing.
+func (thread *Thread) Halt() (err error) {
+	defer func() {
+		if err == nil {
+			thread.running = false
+		}
+	}()
+	if thread.Stopped() {
+		return
+	}
+	err = thread.halt()
+	return
+}
+
 func (t *Thread) halt() (err error) {
 	kret := C.thread_suspend(t.os.threadAct)
 	if kret != C.KERN_SUCCESS {

--- a/pkg/proc/native/threads_linux.go
+++ b/pkg/proc/native/threads_linux.go
@@ -16,15 +16,6 @@ type OSSpecificDetails struct {
 	registers sys.PtraceRegs
 }
 
-// Halt stops this thread from executing.
-func (thread *Thread) Halt() (err error) {
-	if thread.Stopped() {
-		return
-	}
-	err = thread.halt()
-	return
-}
-
 func (t *Thread) halt() (err error) {
 	err = sys.Tgkill(t.dbp.pid, t.ID, sys.SIGSTOP)
 	if err != nil {
@@ -34,7 +25,9 @@ func (t *Thread) halt() (err error) {
 	return
 }
 
-func (t *Thread) stopped() bool {
+// Stopped returns whether the thread is stopped at
+// the operating system level.
+func (t *Thread) Stopped() bool {
 	state := status(t.ID, t.dbp.os.comm)
 	return state == StatusTraceStop || state == StatusTraceStopT
 }

--- a/pkg/proc/native/threads_linux.go
+++ b/pkg/proc/native/threads_linux.go
@@ -16,15 +16,19 @@ type OSSpecificDetails struct {
 	registers sys.PtraceRegs
 }
 
+// Halt stops this thread from executing.
+func (thread *Thread) Halt() (err error) {
+	if thread.Stopped() {
+		return
+	}
+	err = thread.halt()
+	return
+}
+
 func (t *Thread) halt() (err error) {
 	err = sys.Tgkill(t.dbp.pid, t.ID, sys.SIGSTOP)
 	if err != nil {
 		err = fmt.Errorf("halt err %s on thread %d", err, t.ID)
-		return
-	}
-	_, _, err = t.dbp.waitFast(t.ID)
-	if err != nil {
-		err = fmt.Errorf("wait err %s on thread %d", err, t.ID)
 		return
 	}
 	return

--- a/pkg/proc/native/threads_linux.go
+++ b/pkg/proc/native/threads_linux.go
@@ -14,6 +14,7 @@ type WaitStatus sys.WaitStatus
 // process details.
 type OSSpecificDetails struct {
 	registers sys.PtraceRegs
+	running   bool
 }
 
 func (t *Thread) halt() (err error) {
@@ -37,7 +38,7 @@ func (t *Thread) resume() error {
 }
 
 func (t *Thread) resumeWithSig(sig int) (err error) {
-	t.running = true
+	t.os.running = true
 	t.dbp.execPtraceFunc(func() { err = PtraceCont(t.ID, sig) })
 	return
 }

--- a/pkg/proc/native/threads_linux.go
+++ b/pkg/proc/native/threads_linux.go
@@ -17,10 +17,10 @@ type OSSpecificDetails struct {
 	running   bool
 }
 
-func (t *Thread) halt() (err error) {
+func (t *Thread) stop() (err error) {
 	err = sys.Tgkill(t.dbp.pid, t.ID, sys.SIGSTOP)
 	if err != nil {
-		err = fmt.Errorf("halt err %s on thread %d", err, t.ID)
+		err = fmt.Errorf("stop err %s on thread %d", err, t.ID)
 		return
 	}
 	return

--- a/pkg/proc/native/threads_windows.go
+++ b/pkg/proc/native/threads_windows.go
@@ -18,6 +18,20 @@ type OSSpecificDetails struct {
 	hThread syscall.Handle
 }
 
+// Halt stops this thread from executing.
+func (thread *Thread) Halt() (err error) {
+	defer func() {
+		if err == nil {
+			thread.running = false
+		}
+	}()
+	if thread.Stopped() {
+		return
+	}
+	err = thread.halt()
+	return
+}
+
 func (t *Thread) halt() (err error) {
 	// Ignore the request to halt. On Windows, all threads are halted
 	// on return from WaitForDebugEvent.

--- a/pkg/proc/native/threads_windows.go
+++ b/pkg/proc/native/threads_windows.go
@@ -18,30 +18,6 @@ type OSSpecificDetails struct {
 	hThread syscall.Handle
 }
 
-// Halt stops this thread from executing.
-func (thread *Thread) Halt() (err error) {
-	defer func() {
-		if err == nil {
-			thread.running = false
-		}
-	}()
-	if thread.Stopped() {
-		return
-	}
-	err = thread.halt()
-	return
-}
-
-func (t *Thread) halt() (err error) {
-	// Ignore the request to halt. On Windows, all threads are halted
-	// on return from WaitForDebugEvent.
-	return nil
-
-	// TODO - This may not be correct in all usages of dbp.Halt.  There
-	// are some callers who use dbp.Halt() to stop the process when it is not
-	// already broken on a debug event.
-}
-
 func (t *Thread) singleStep() error {
 	context := newCONTEXT()
 	context.ContextFlags = _CONTEXT_ALL
@@ -140,9 +116,9 @@ func (t *Thread) Blocked() bool {
 	}
 }
 
-func (t *Thread) stopped() bool {
-	// TODO: We are assuming that threads are always stopped
-	// during command execution.
+// Stopped returns whether the thread is stopped at the operating system
+// level. On windows this always returns true.
+func (t *Thread) Stopped() bool {
 	return true
 }
 

--- a/pkg/proc/native/threads_windows.go
+++ b/pkg/proc/native/threads_windows.go
@@ -86,7 +86,6 @@ func (t *Thread) singleStep() error {
 }
 
 func (t *Thread) resume() error {
-	t.running = true
 	var err error
 	t.dbp.execPtraceFunc(func() {
 		//TODO: Note that we are ignoring the thread we were asked to continue and are continuing the

--- a/pkg/proc/proc_test.go
+++ b/pkg/proc/proc_test.go
@@ -3410,3 +3410,45 @@ func TestIssue1137(t *testing.T) {
 		assertNoError(v2.Unreadable, t, "iface2 unreadable")
 	})
 }
+
+func TestIssue1101(t *testing.T) {
+	// If a breakpoint is hit close to process death on a thread that isn't the
+	// group leader the process could die while we are trying to stop it.
+	//
+	// This can be easily reproduced by having the goroutine that's executing
+	// main.main (which will almost always run on the thread group leader) wait
+	// for a second goroutine before exiting, then setting a breakpoint on the
+	// second goroutine and stepping through it (see TestIssue1101 in
+	// proc_test.go).
+	//
+	// When stepping over the return instruction of main.f the deferred
+	// wg.Done() call will be executed which will cause the main goroutine to
+	// resume and proceed to exit. Both the temporary breakpoint on wg.Done and
+	// the temporary breakpoint on the return address of main.f will be in
+	// close proximity to main.main calling os.Exit() and causing the death of
+	// the thread group leader.
+
+	withTestProcess("issue1101", t, func(p proc.Process, fixture protest.Fixture) {
+		_, err := setFunctionBreakpoint(p, "main.f")
+		assertNoError(err, t, "setFunctionBreakpoint()")
+		assertNoError(proc.Continue(p), t, "Continue()")
+		assertNoError(proc.Next(p), t, "Next() 1")
+		assertNoError(proc.Next(p), t, "Next() 2")
+		lastCmd := "Next() 3"
+		exitErr := proc.Next(p)
+		if exitErr == nil {
+			lastCmd = "final Continue()"
+			exitErr = proc.Continue(p)
+		}
+		if pexit, exited := exitErr.(proc.ProcessExitedError); exited {
+			if pexit.Status != 2 && testBackend != "lldb" {
+				// looks like there's a bug with debugserver on macOS that sometimes
+				// will report exit status 0 instead of the proper exit status.
+				t.Fatalf("process exited status %d (expected 2)", pexit.Status)
+			}
+		} else {
+			assertNoError(exitErr, t, lastCmd)
+			t.Fatalf("process did not exit after %s", lastCmd)
+		}
+	})
+}

--- a/pkg/proc/proc_test.go
+++ b/pkg/proc/proc_test.go
@@ -75,7 +75,6 @@ func withTestProcessArgs(name string, t testing.TB, wd string, args []string, bu
 	}
 
 	defer func() {
-		p.Halt()
 		p.Detach(true)
 		if tracedir != "" {
 			protest.SafeRemoveAll(tracedir)

--- a/pkg/proc/proc_test.go
+++ b/pkg/proc/proc_test.go
@@ -969,20 +969,6 @@ func TestKill(t *testing.T) {
 		return
 	}
 	withTestProcess("testprog", t, func(p proc.Process, fixture protest.Fixture) {
-		if err := p.Kill(); err != nil {
-			t.Fatal(err)
-		}
-		if !p.Exited() {
-			t.Fatal("expected process to have exited")
-		}
-		if runtime.GOOS == "linux" {
-			_, err := os.Open(fmt.Sprintf("/proc/%d/", p.Pid()))
-			if err == nil {
-				t.Fatal("process has not exited", p.Pid())
-			}
-		}
-	})
-	withTestProcess("testprog", t, func(p proc.Process, fixture protest.Fixture) {
 		if err := p.Detach(true); err != nil {
 			t.Fatal(err)
 		}
@@ -2086,8 +2072,7 @@ func TestUnsupportedArch(t *testing.T) {
 	case proc.UnsupportedLinuxArchErr, proc.UnsupportedWindowsArchErr, proc.UnsupportedDarwinArchErr:
 		// all good
 	case nil:
-		p.Halt()
-		p.Kill()
+		p.Detach(true)
 		t.Fatal("Launch is expected to fail, but succeeded")
 	default:
 		t.Fatal(err)

--- a/pkg/proc/proc_unix_test.go
+++ b/pkg/proc/proc_unix_test.go
@@ -46,7 +46,7 @@ func TestIssue419(t *testing.T) {
 			if p.Pid() <= 0 {
 				// if we don't stop the inferior the test will never finish
 				p.RequestManualStop()
-				err := p.Kill()
+				err := p.Detach(true)
 				errChan <- errIssue419{pid: p.Pid(), err: err}
 				return
 			}
@@ -57,7 +57,7 @@ func TestIssue419(t *testing.T) {
 		errChan <- proc.Continue(p)
 	})
 
-	for i :=0; i<2; i++ {
+	for i := 0; i < 2; i++ {
 		err := <-errChan
 
 		if v, ok := err.(errIssue419); ok {

--- a/service/test/variables_test.go
+++ b/service/test/variables_test.go
@@ -128,7 +128,6 @@ func withTestProcess(name string, t *testing.T, fn func(p proc.Process, fixture 
 	}
 
 	defer func() {
-		p.Halt()
 		p.Detach(true)
 		if tracedir != "" {
 			protest.SafeRemoveAll(tracedir)

